### ISSUE TITLE
Ordered factors break DataFrame constructor

### ISSUE THREAD

**Issue by PeteHaitch** [Bioconductor/S4Vectors#23]
``` r
suppressPackageStartupMessages(library(S4Vectors))
DataFrame(x = factor(letters),
          y = factor(LETTERS, ordered = TRUE))
#> Error in DataFrame(x = factor(letters), y = factor(LETTERS, ordered = TRUE)): cannot coerce class "orderedfactor" to a DataFrame
# Works when first wrapped in a data.frame
DataFrame(data.frame(x = factor(letters),
                     y = factor(LETTERS, ordered = TRUE)))
#> DataFrame with 26 rows and 2 columns
#>            x         y
#>     <factor> <ordered>
#> 1          a         A
#> 2          b         B
#> 3          c         C
#> 4          d         D
#> 5          e         E
#> ...      ...       ...
#> 22         v         V
#> 23         w         W
#> 24         x         X
#> 25         y         Y
#> 26         z         Z
```

<sup>Created on 2018-10-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.1 (2018-07-02)
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language en_AU:en                    
#>  collate  en_AU.UTF-8                 
#>  tz       Australia/Melbourne         
#>  date     2018-10-08
#> Packages -----------------------------------------------------------------
#>  package      * version date       source        
#>  backports      1.1.2   2017-12-13 CRAN (R 3.5.0)
#>  base         * 3.5.1   2018-07-03 local         
#>  BiocGenerics * 0.27.1  2018-06-17 Bioconductor  
#>  compiler       3.5.1   2018-07-03 local         
#>  datasets     * 3.5.1   2018-07-03 local         
#>  devtools       1.13.6  2018-06-27 CRAN (R 3.5.1)
#>  digest         0.6.17  2018-09-12 CRAN (R 3.5.1)
#>  evaluate       0.11    2018-07-17 CRAN (R 3.5.1)
#>  graphics     * 3.5.1   2018-07-03 local         
#>  grDevices    * 3.5.1   2018-07-03 local         
#>  htmltools      0.3.6   2017-04-28 CRAN (R 3.5.0)
#>  knitr          1.20    2018-02-20 CRAN (R 3.5.0)
#>  magrittr       1.5     2014-11-22 CRAN (R 3.5.0)
#>  memoise        1.1.0   2017-04-21 CRAN (R 3.5.1)
#>  methods      * 3.5.1   2018-07-03 local         
#>  parallel     * 3.5.1   2018-07-03 local         
#>  Rcpp           0.12.18 2018-07-23 CRAN (R 3.5.1)
#>  rmarkdown      1.10    2018-06-11 CRAN (R 3.5.0)
#>  rprojroot      1.3-2   2018-01-03 CRAN (R 3.5.1)
#>  S4Vectors    * 0.19.20 2018-09-29 Bioconductor  
#>  stats        * 3.5.1   2018-07-03 local         
#>  stats4       * 3.5.1   2018-07-03 local         
#>  stringi        1.2.4   2018-07-20 CRAN (R 3.5.1)
#>  stringr        1.3.1   2018-05-10 CRAN (R 3.5.0)
#>  tools          3.5.1   2018-07-03 local         
#>  utils        * 3.5.1   2018-07-03 local         
#>  withr          2.1.2   2018-03-15 CRAN (R 3.5.0)
#>  yaml           2.2.0   2018-07-25 CRAN (R 3.5.1)
```

</details>

**Comment by lawremi** [comment]
Yes and fairly specifically in the cited example. Here is a more direct one:
```
library(S4Vectors)
as(factor(letters), "DataFrame") # necessary due to caching side effect
as(factor(letters, ordered=TRUE), "DataFrame")
# Error in if (any(diffs < 0)) { : missing value where TRUE/FALSE needed
```

Looks like dispatching gets confused due to the inconsistent class hierarchy between "factor" and "ordered". The "ordered" class is defined simply via
```
setOldClass(c("ordered", "factor"))
```
while the "factor" class gets an S4 definition (simplified below):
```
setClass("factor", contains = "integer", representation(levels = "character"))
setOldClass("factor", S4Class = "factor")
```

It is a bit confusing that a rich S4 class is derived from "oldClass".  There are three classes that do this: "ts", "factor" and "data.frame". The "anova" class has a similar problem to "ordered".  Any S3 object that extends one of those classes will probably want to have the inheritance carried over onto its corresponding old class by defining an actual S4 class and then passing it via `S4Class=`.

After a fix to `setOldClass()`, it looks like we can define "ordered" correctly. I've also changed the dispatch logic so that it treats this case as a cache miss and falls back gracefully. Finally, S4Vectors 0.19.21 not only fixes the broken error message above ("orderedfactor") but also might avoid this case entirely by removing a hopefully unnecessary method.

**Comment by PeteHaitch** [comment]
Thanks, Michael. Fixed in v0.19.22
